pkg/util/resource.go

```diff
@@ -0,0 +1,92 @@
+package util
+
+import corev1 "k8s.io/api/core/v1"
+
+// Resource is a collection of compute resource.
+type Resource struct {
+	MilliCPU int64
+	Memory   int64
+}
+
+// EmptyResource creates a empty resource object and returns.
+func EmptyResource() *Resource {
+	return &Resource{}
+}
+
+// NewResource creates a new resource object from resource list.
+func NewResource(rl corev1.ResourceList) *Resource {
+	r := EmptyResource()
+	for rName, rQuant := range rl {
+		switch rName {
+		case corev1.ResourceCPU:
+			r.MilliCPU += rQuant.MilliValue()
+		case corev1.ResourceMemory:
+			r.Memory += rQuant.Value()
+		default:
+			continue
+		}
+	}
+	return r
+}
+
+// AddResource is used to add two resources.
+func (r *Resource) AddResource(rl corev1.ResourceList) {
+	if r == nil {
+		return
+	}
+
+	for rName, rQuant := range rl {
+		switch rName {
+		case corev1.ResourceCPU:
+			r.MilliCPU += rQuant.MilliValue()
+		case corev1.ResourceMemory:
+			r.Memory += rQuant.Value()
+		default:
+			continue
+		}
+	}
+}
+
+// SetMaxResource compares with ResourceList and takes max value for each Resource.
+func (r *Resource) SetMaxResource(rl corev1.ResourceList) {
+	if r == nil {
+		return
+	}
+
+	for rName, rQuant := range rl {
+		switch rName {
+		case corev1.ResourceCPU:
+			if r.MilliCPU < rQuant.MilliValue() {
+				r.MilliCPU = rQuant.MilliValue()
+			}
+		case corev1.ResourceMemory:
+			if r.Memory < rQuant.Value() {
+				r.Memory = rQuant.Value()
+			}
+		default:
+			continue
+		}
+	}
+}
+
+// GetPodResourceWithoutInitContainers returns Pod's resource request, it does not contain
+// init containers' resource request.
+func GetPodResourceWithoutInitContainers(pod *corev1.Pod) *Resource {
+	result := EmptyResource()
+	for _, container := range pod.Spec.Containers {
+		result.AddResource(container.Resources.Requests)
+	}
+
+	return result
+}
+
+// GetPodResourceRequest returns all the resource required for that pod.
+func GetPodResourceRequest(pod *corev1.Pod) *Resource {
```

Comment:
It will be wonderful to give more comments here, I guess not all guys know the rules.
How about this:
```
// CalculateRequestResource calculates the resource required for pod, that is the higher of:
// - the sum of all app containers(spec.Containers) request for a resource.
// - the effective init containers(spec.InitContainers) request for a resource.
// The effective init containers request is the highest request on all init containers.
func CalculateRequestResource(pod *corev1.Pod) *Resource {
```

Reply:
That's good.